create function "years_select" which shows data according to input year

DarthProvader opened the pull request.

DarthProvader committed:
create function "years_select" which shows data according to year

OpposedOak reviewed (dismissed):
Approved

DarthProvader committed:
uncommented some prints

DarthProvader committed:
merged with main

DarthProvader committed:
Daily average funtion created

Function get_daily_average_temperatures created

DarthProvader committed:
create get yearly temperatures function

DarthProvader committed:
created separate file for api

DarthProvader committed:
file process_meteo_data up to date with main

DarthProvader committed:
Added function

get_daily_average_temperatures changed so it takes correct date format
get_yearly_average_temperatures created - returns disctionary of average temperatures yearly

zitkat commented:
Ahoj, navrhuji, abyste si mergnuli main s tím že změny v process_meteo_data.py přijmete z main, api_meteo_data.py necháte tak jak je. To je asi nejsnažší udělat přes nějaké IDE, kdybste měli čas můžeme to rychle udělat spolu, v nejhorším to můžu udělat i já. Potřebné věci pak naimportujte z process_meteo_data.py.

DarthProvader commented:
Ahoj, 
Dneska v 6 máme s Vaškem v plánu na to kouknout. Pokud máš čas, tak by ses mohl přidat a uděláme to společně.

zitkat commented:
> Dneska v 6 máme s Vaškem v plánu na to kouknout. Pokud máš čas, tak by ses mohl přidat a uděláme to společně.

Ok, tak jo, tak mi pak akorát prosím napište nějakou soukromou zprávu (mail, FB, ...) jak se spojíme :-)

DarthProvader commented:
Přidal jsem si tě na fb. Většinou jsme na https://meet.vpsfree.cz/pythonplzen ale to případně ještě potvrdim

Mintaka commented:
Vypadá to dobře.
Teď si jen dát tu práci to přeskládat do současného main.
Jaký @zitkat navrhuješ způsob?

DarthProvader committed:
Merge branch 'main' into issue_8

DarthProvader committed:
setup_db_connection import from process_meteo_data and add main

DarthProvader commented:
> Vypadá to dobře.
> Teď si jen dát tu práci to přeskládat do současného main.
> Jaký @zitkat navrhuješ způsob?

Už jsme to s Tomášem vyřešili. Teď už process_meteo_data vypadá stejně jako v main a náš soubor je oddělen takže si vždycky můžeme udělat git pull a api bude pracovat s aktuální databází/programem.

OpposedOak committed:
Update api_meteo_data.py

Function "get_monthly_average " has been added

OpposedOak committed:
Update api_meteo_data.py

Bug Fix in for loop  "for k,v in year_month.items(): "in function "get_monthly_average_temperatures"

OpposedOak committed:
Update api_meteo_data.py

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+
+
+
+#date: YYYY-MM-DD
```
Tohle je důležitá informace, která by měla být v dokumentačním komentáři -- string se třemi uvozovkami pod názvem funkce, třeba takhle:
```python
def get_daily_average_temperatures(cursor, date_from=None, date_to=None, meteostations=None, limit=None):
     """
     This function just bla bla bla...
     :param date_from: YYYY-MM-DD
     :param date_to: YYYY-MM-DD
     """
     ...
```
Takto to umí leckterá IDE zpracovat a zobrazit jako nápovědu, existuje několik formátů jak dokumentaci psát, mohli bychom nějaký vybrat a dohodnout se na zásadách. Zatím zkuste doplnit stručnou dokumentaci k téhle funkci a zjistit, jestli to vaše IDE dokáže využít, uvidíte, že to může být užitečné.

DarthProvader replied:
Done ce5f434

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+    count_years = date_from_y
+    for rok in range((date_to_y - date_from_y) + 1): #create list called "years" which consist of range of years from selected dates (eg 1995-05-4 and 1998-04-09 creates years[1995, 1996, 1997, 1998])
+        years.append(count_years)
+        count_years += 1
```
Proč generovat roky, které chceme zprůměrovat takhle složitě na vytváření posloupností máme přece v pythonu funkci ne? Tou půjde zjednodušit i kód níže ;-) 
Tip: kdykoliv máte ve for cyklu jen tak volně výraz 
```python
něco += 1
```
je to podezdřelé.

DarthProvader replied:
Mohl bych se zeptat jakou funkci máš na mysli? Generovat roky chci abych si udělal seznam roků a nad každým udělal query. Jak jinak by se to dalo provést?

zitkat replied:
Asi jsem měl napovědět trochu jasněji, mám na mysli range, proč nepoužít range ke generování roků přímo místo všeho toho počítání?

DarthProvader replied:
Tak snad by to takhle mělo být lepší
c644be9

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+        WHERE date LIKE '{years[i]}%'""")
+        avg_temp = str(query.fetchall())
+        results[f"{years[i]}"] = avg_temp[2:5] #for every year get average temperature and put it into the dictionary as string
+        i += 1
```
Proč se v cyklu nikde nevyužívá proměnná rok? Co se do ní v průběhu cyklu ukládá? Je nutné převádět proměnnou `year` na řetězec pro indexování do slovníku?

DarthProvader replied:
Opraveno v ce5f434
Nyní je type(avg_temp) = float, místo "i" jsem použil pro loop proměnnou "rok".
Omezeno na 2 desetinná místa.

zitkat replied:
`float` je ok, proč se ale omezovat na dvě desetinná místa? Obdobně jako výše, proč generovat roky dopředu, když si je můžu vyrobit pomocí range a ušetřit si tak dokonce dvě proměnné a indexování do seznamu years?

DarthProvader replied:
c644be9
viz výše + odstraněny restrikce na 2 des. čísla

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+        for i in v.keys(): 
+            query = cursor.execute(f"""SELECT Avg(temperature) FROM temperatures
+                                    WHERE date LIKE '{k}-%{i}%'""")
+            avg_temp = str(query.fetchall())
+            results_month [i] = avg_temp[2:7]
```
Funkce vracející průměry by neměla vracet řetězce.

DarthProvader replied:
ce5f434
Měl by rok být taky integer?

zitkat replied:
Asi jo, rozhodně se s nimi bude později lépe pracovat.

DarthProvader replied:
c644be9

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+        
+        year_actuall += 1
+
+    for k,v in year_month.items():
```
`k` a `v` nejsou vhodné názvy proměnných.

OpposedOak replied:
Přejmenuji. Jsou to zatím jen pracovní názvy.

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+           months = {}
+           count_months = date_from_m
+           for month in range(date_from_m,date_to_m + 1):
+               months [month] = "NaN"
```
Pokud se hranaté závorky používají k indexaci nedělá se před nimi mezera, podobně jako při volání funkce.

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+           months = {}
+           count_months = date_from_m
+           for month in range(date_from_m,date_to_m + 1):
+               months [month] = "NaN"
```
Do slovníku months se uloží jen samé `"NaN"`, nikdy žádná užitečná data a požívají se jen klíče, není k tomu lepší seznam?

OpposedOak replied:
Řetežec "NaN" funguje jako place-holder, později se na jeho místo přiřadí teplota z databáze. Šlo by to udělat jednodušeji ?

zitkat replied:
Ať koukám jak koukám, tak nikde nevidím, že by se tam něco později zapisovalo. Na konci funkce jsou v ní pořad jen ty `NaNy`

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+           count_months = date_from_m
+           for month in range(date_from_m,date_to_m + 1):
+               months [month] = "NaN"
+               count_months += 1
```
K čemu je proměnná `count_months`její hodnota se nikde nepoužije.

OpposedOak replied:
Proměnná vyřazena

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+    date_to_m = int(list(date_to.split("-"))[1])
+
+    year_month = {}
+    results_month = {}
```
Proměnná `results_months` se nikde nepoužije.

OpposedOak replied:
Používá se na přiřazování hodnot do slovníku results:  "results [k] = results_month". Určitě bude existovat lepší způsob

zitkat replied:
Už to vidím, inicializuje se při každém průchodu cyklem 
```python
for k,v in year_month.items():
```
takže to z pohledu řádku 45 vypadá, že se nepoužívá. Tuhle inicializaci můžete smazat, to už je ale drobnost :-)

zitkat commented on api_meteo_data.py:
```diff
@@ -0,0 +1,98 @@
+    results = {}
+    year_actuall = date_from_y
+
+    while year_actuall != date_to_y + 1:
+        
+        if date_from_y == date_to_y:
+           months = {}
+           count_months = date_from_m
+           for month in range(date_from_m,date_to_m + 1):
+               months [month] = "NaN"
+               count_months += 1
+           year_month [year_actuall] = months 
+        
+        if (year_actuall == date_from_y) and date_from_y != date_to_y:
+            months = {}
+            count_months = date_from_m
+            for month in range(date_from_m,13):
+                months [month] = "NaN"
+                count_months += 1
+            year_month [year_actuall] = months
+                
+        elif year_actuall == date_to_y:
+            months = {}
+            count_months = 1
+            for month in range(1,date_to_m + 1):
+                months [month] = "NaN"
+                count_months += 1
+            year_month [year_actuall] = months
+            
+        else:
+            months = {}
+            for month in range(1,13):
+                months [month] = "NaN"             
+            year_month [year_actuall] = months
+        
+        year_actuall += 1
```
Tenhle kód je dost dlouhý, i po přípomínkách výše nejspíš bude, hodilo by se udělat z něj funkci, napadá mě i jméno, `generate_years_months`.

OpposedOak replied:
Do funkce ho můžeme přesunutou, pokud budeme zachovávat způsob sestavení slovníků pro výběr teplot

OpposedOak committed:
Update api_meteo_data.py

New approach for function  "get_weekly_average_temperatures"

DarthProvader committed:
get daily average extended for meteonstations query

zitkat reviewed (commented):
Aha, tak jsem zjistil, že pokud člověk nenapíše ještě nějaké milé úvodní slovo tak se review nezobrazuje a já se divím, že nikdo nereaguje :-(

DarthProvader committed:
Update api

Function explanations added, meteostation param is taken into consideration and some loops updated

DarthProvader committed:
Additional fixes

get_yearly_data simplified loops and now returns dictionary as {int(year):float(avg_temp)}
No restrictions for decimals in avg temp

OpposedOak committed:
Update api_meteo_data.py

Data are grouped by week

OpposedOak committed:
Update api_meteo_data.py

OpposedOak committed:
Update api_meteo_data.py

OpposedOak committed:
Updated function get weekly average

OpposedOak committed:
Update api_meteo_data.py
